### PR TITLE
Sketching out SAMSTS structure - not operational yet

### PR DESCRIPTION
**Pull request by max-nova**

### CODE REVIEW

File: samsts/stack.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    """
+    BANDS_PER_SCENE = 5
+
+    def __init__(self, landsat_fns):
+        """
+        The Landsat filenames need to be the scene id.  Scenes should already
+        be TOAR'd, have pixels aligned, and have clouds masked out.  The scenes
+        should all have the same extent.
+        """
+        date_to_scene = dict((parse_date(fn), fn) for fn in landsat_fns)
```
Comment (mholkesvik) [inline samsts/stack.py]: not seeing the parse_date method?

File: samsts/cluster.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    clusters = kmeans_cluster(segments, obs50)
+    return merge_clusters(clusters, obs50)
+
+
+def kmeans_cluster(segments, obs50, clusters=[], samr_thresh=SAMR_THRESH_INIT):
+    """
+    Cluster the segments using a K-means inspired approach.
+    Returns a list of clusters.
+
+    (Section 5.2.2, paragraph 1)
```
Comment (mholkesvik) [inline samsts/cluster.py]: any way we can use http://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html  ????

File: samsts/cluster.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+    def merge(self, other):
+        self.segments += other.segments
+        self.update_center()
+        self.initialize()
+
+
+def cluster(segments, obs50):
+    """
+    Given a list of segments
```
Comment (mholkesvik) [inline samsts/cluster.py]: fragment?

also, I'm generally not wild about the "obs50" variable - is there something better we could use?

File: samsts/cluster.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                c2._processed = True
+                merged_clusters.append(c1.merge(c2))
+        else:
+            merged_clusters.append(c1)
+
+    end_num_clusters = len(merged_clusters)
+    if end_num_clusters == start_num_clusters:
+        return merged_clusters
+    else:
+        return merge_clusters(merged_clusters, obs50, iter_num+1)
```
Comment (mholkesvik) [inline samsts/cluster.py]: sneaky sneaky

File: samsts/segment.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+
+def fill_segment_gaps(scene, segs):
+    """
+    Given a particular scene and a list of segments that have had their closest
+    clusters set, fill in the gap pixels
+    """
+    for s in segs:
+        if not s.clusters:
+            raise Exception('No clusters set on segment - filling aborted')
```
Comment (mholkesvik) [inline samsts/segment.py]: assert s.clusters, 'No clusters set on segment - filling aborted'

File: samsts/run.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    small_clusters = cluster.cluster(small_segs, obs50)
+
+    # associate 10 closest clusters with each segment
+    for s in big_segs:
+        s.set_clusters(s.closest_clusters(big_clusters))
+    for s in small_segs:
+        s.set_cluster(s.closest_clusters(small_clusters))
+
+    for scene in scenes:
+        segment.fill_segment_gaps(scene, big_segs)
```
Comment (mholkesvik) [inline samsts/run.py]: multicore at scene level or segment level? Both?

File: samsts/sam.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import numpy as np
+
+
+def calc_samr(x, y, obs50):
+    """
+    Compute the refined spectral angle mapper similarity metric (SAMr) which
+    corrects for cases in which there is a high proportion of missing
+    observations (Eqn 2).
+
+    Takes in two numpy arrays, x and y, and returns a scalar.
```
Comment (mholkesvik) [inline samsts/sam.py]: is this function generic to 3 and 2D numpy arrays? if not, will the scalar always be a float? why not call it that?

File: samsts/stack.py (hunk 1 above)
Comment (max-nova) [reply samsts/stack.py]: Doesn't exist yet

File: samsts/cluster.py (hunk 2 above)
Comment (max-nova) [reply samsts/cluster.py]: No, because his algorithm is only k-means "inspired" and not really k-means.

File: samsts/cluster.py (hunk 3 above)
Comment (max-nova) [reply samsts/cluster.py]: That's what he called it in the paper.  I'm also a bit torn between using the variable names from the paper (for easy cross-referencing) and names that actually make sense.

File: samsts/segment.py (hunk 5 above)
Comment (max-nova) [reply samsts/segment.py]: Ah, better move

File: samsts/run.py (hunk 6 above)
Comment (max-nova) [reply samsts/run.py]: To discuss.  Let's see how the computational profile shakes out

File: samsts/sam.py (hunk 7 above)
Comment (max-nova) [reply samsts/sam.py]: not sure I understand the comment.  This should really only run on 1D arrays.

File: samsts/segment.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+
+PIX_SAM_MERGE_THRESH = 0.9995  # from paper (Section 5.2.1)
+SEG_MERGE_MAX_ITERS = 3  # from the paper (Section 5.2.1)
+SEG_SMALL_SIZE_THRESH = 3  # from the paper (Section 5.2.1)
+CLOSEST_CLUSTER_NUM = 10  # "M" from the paper (Section 5.2.2)
+
+
+class Segment:
+    """
+    Represents a contiguous chunk of pixels that are similar to each other
```
[the remark this answers is not in the document]
Comment (max-nova) [reply samsts/segment.py]: they're 2D, yes.

File: samsts/segment.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+        columns
+        """
+        return self.stack.get_values(self.mask)
+
+    def has_gaps(self, scene):
+        """
+        Return True if there are any "gap" pixels in the segment in the
+        specified scene
+        """
+        return any(np.isnan(
```
[the remark this answers is not in the document]
Comment (max-nova) [reply samsts/segment.py]: One of the things we need to be careful about.  A plumbing detail

**Comment by mholkesvik** [comment]
not most recent branch

**Pull request closed without merging**